### PR TITLE
Updates to make user-side flux (and near2far) computation consistent

### PR DESCRIPTION
**Pull request by momchil-flex**
The fundamental issue that this PR addresses is the following. Say you have a primal grid along `z` defined on `n dz` and a dual grid defined on `(n + 0.5) dz`. Say also that you have a perfectly injected mode propagating along z, such that the frequency-domain tangential fields on the primal grid are exactly given by some `E0 exp(1j * k * n * dz)`, `H0 exp(1j * k * (n + 0.5) * dz)`, where k is the effective wave vector of the mode at a particular frequency. Now, if you pick an arbitrary position `z` along the propagation direction, and use linear interpolation to compute `E(z)` and `H(z)`, and then use those to compute the flux at that location, you will get values that *oscillate* with `z` weakly but non-negligibly.

So far, to hide this from the user, we have been normalizing the sources to inject `flux = 1` when computed *on the dual grid*, i.e. at the cell centers. Note that the difference between this flux and the analytic one coming from E0 and H0 is a factor of `cos(k * dz / 2)`. This factor is specific to interpolating to a grid center, and comes entirely from the E-field interpolation, as the H fields are not interpolated as they are defined exactly at the center. This factor is what would vary if you were to vary the interpolation position. To avoid this, and to match the source normalization, the backend `FluxMonitor`-s have also been always returning the flux at the grid centers, and similarly that's how mode amps are computed, such that in the end `flux` and `np.abs(amps) ** 2` are always very close to 1.

However, we have introduced `FieldData.flux` and `FieldData.dot` methods available to the user, which can be applied to any 2D field data where the fields have been interpolated to an arbitrary `z`. This results in the unphysical oscillations of the flux as discusses above, and also makes user-side and server-side results differ if a `FieldMonitor` is not placed exactly at a cell center.

There are two main stipulations of this PR:
- The first is that the correct way to compute flux along the *normal* direction is *not* to directly interpolate E(z) and H(z) and then compute the flux, but rather to compute the flux at the grid centers surrounding z (by interpolating the fields to those centers), and then interpolate *the flux* to z. This ensures that in the example above, the flux will vary very weakly with `z`, and, due to the source normalization to inject unit power as recorded at the *centers*, the flux will be very close to 1 regardless of `z`. This is now what backend `FluxMonitor`-s do, and they now create an illusion of continuity along all 3 directions. Along the tangential directions, the computation is much more straightforward: we interpolate fields to centers, compute the Poynting vector, and integrate over the monitor area, adjusting the differential area of the cells at the edges to match the exact monitor dimensions.
- The second is that when snapping 2D `FieldMonitor` results to the exact location along `z`, the first consideration needs to be taken into account. So now, instead of directly interpolating E(z) and H(z) and be done with it, we do something more complicated. On the backend, we expand the `FieldMonitor` sufficiently along `z` to be able to compute the flux at the two cell centers surrounding the exact monitor `z`. Then, we compute the flux at `z` in two different ways:
  -  By interpolating E and H to z (call this `flux_interp1`). This is what `FieldData.flux` would return for that monitor if we don't do anything else.
  -  By interpolating E and H to `(n + 0.5) dz` and `(n + 1.5) dz` (center positions around `z`), computing the flux, and then interpolating *that* to `z` (call this `flux_interp2`).

   Finally, we scale all the fields of the `FieldMonitor` by `np.sqrt(flux_interp2 / flux_interp1)`. Now, `FieldData.flux` for this monitor will return the desired `flux_interp2`, and match the server-side results from a `FluxMonitor`.

Similar considerations need to be applied to make user-side and server-side mode amps computed as `ModeSolverData.dot(FieldData)` consistent. However, since the mode solver does not take the grid along the normal direction into account, we cannot really automatically apply the correction in that case (consider for example the fact that the `ModeSolverData` could be coming from a local mode solver run). Well, we *could*, but it would mean that the mode solver results would depend on the grid along the normal direction, and even more unexpectedly, on the exact placement of the `ModeSolverMonitor` along that direction with respect to the Yee grid. To avoid this, I have instead introduced a `ModeSolverData.finite_grid_copy(grid)` method. Thus the user can replicate server-side mode amplitudes by doing e.g. `ModeSolverData.finite_grid_copy(sim.grid).dot(FieldData)`.

These changes, together with the corresponding backend ones, have the following nice features:
- User-side and server-side flux and amps computations should now be always equivalent.
- Flux monitors (and the `flux` method of field monitors) take the analytic geometry of the monitor into account as well as possible, under the constraints of doing a linear interpolation on the Yee grid.

And here are some caveats *introduced* by the new handling:
- Pretty much all monitors should be placed at least *two* cells away from a source or structure interface to avoid numerical artifacts. Should we put this somewhere in the docs, and how? Should we have it as a note under every monitor?
- If someone puts a 3D `FieldMonitor` and interpolates the fields to a given z, they will not get the same results as a 2D `FieldMonitor` at that z (unless it's a grid cell center). We could again explain this in a note in the monitors docstrings?

Some other updates:
- The `flux` and `dot` methods now also handle monitors with `colocate=True` correctly. For down-sampling monitors, they raise a not implemented error. I also added trivial `colocate` and `interval_space` properties to `ModeSolverData` to unify this treatment.
- I've removed `ModeSolverDataset.sel_mode_index` since it's a bit inconsistent and our strategy tends to be to *not* implement methods like this.
- I've introduced an extra kwarg "expansion" kwarg in the `discretize` methods that's related to the flux interpolation along the normal direction discussed above.
- Added a `ElectromagneticFieldData.time_reversed_copy` property that can be used to compute mode amplitudes in the backward direction. Note: for time-domain data, this just flips the sign of `H`. Does this make sense? If the data was a regular numpy array, I could be reversing the indexing along the time dimension. However, in the xarray data the coords are labeled, so it doesn't seem necessary?

### CODE REVIEW

**Review by shashwat-sh — APPROVED** [review]

**Comment by momchil-flex** [comment]
> Looks good, my comments are mostly just for my own understanding:
> 
> 1. > This ensures that in the example above, the flux will vary very weakly with z
>    
>    
>    So in the new approach, that `cos(k * dz / 2)` factor seems to become instead a linear function of `dz` due to the interpolation - is that the very weak variation you mention?

Actually, for an arbitrary `z` you need to interpolate both E and H so the factor is not linear (it's the product of the two), and each is a bit more complicated than a simple cos. But anyway, let me rephrase my statement: in my example, if the fields are truly given by `E0 exp(1j k z)` and `H0 exp(1j k z)`, the flux will actually not vary *at all* with `z` in this new handling. This is because 1. the source normalization already applies the factor `cos(k * dz / 2)` which ensures that the flux recorded at all cell centers will be unity, and 2. we interpolate the values in two consecutive cell centers, which are unity. I said that it "will vary very weakly" because numerically you do get a tiny bit of variation of the fields as they propagate, but this is not coming from the interpolation anymore and is much smaller than the interpolation contribution if not handled right.

> 2. > Pretty much all monitors should be placed at least two cells away from a source or structure interface
>    
>    
>    
>    * For the case of a structure with a monitor nearby, is the concern that you'd have to then take the permittivity difference into account for the extended points which lie in a different material? I.e. this would mean that the rescaled fields would not be correct?
>    * A monitor defined very close to a structure seems like a situation that can show up quite easily - maybe in addition to the docs, there could be a validator that warns about this? This is tricky though, because it needs knowledge of the grid...might be one of those "pre-upload" things...
>    * I'm also wondering about the case where a user wants the fields very close to a structure's surface but doesn't care about the flux (I imagine this might come up more with 2D materials, impedance boundary sheets, etc., or maybe if the user wants to sample fields in one sim and use them as a custom source in another, I can imagine those fields being sampled at the interface between "region to be modeled" and "region to be modeled separately"...but maybe this situation is too contrived and won't actually happen?). I guess in these cases, one could use a volume monitor and then interpolate within it? Maybe that could be stated as an alternative to a 2D monitor wherever you decide to mention this caveat?
>    * Another option could be an `apply_flux_correction` field in the monitor (default `True`), which enables the backend field rescaling but introduces the restriction of placing the monitor 2 cells away. If the user sets it to `False` and then later calls `.flux`, a warning could be issued saying that the flux may not be accurate, please enable `apply_flux_correction`?

You are right. I am trying to figure out if my statement actually applies to structure interfaces or if it only applies to sources, in order to figure out what is the best way to handle this. The new procedure certainly produces *different* results around interfaces. Say you have something like this:
```
z0                z1                z2                 z3
 |      z0.5       |       z1.5      |      z2.5     |
               ^                    ^
          z_mnt          z_interface
```
where `z_mnt` is the monitor position, `z_interface` is the position of an interface where e.g. most of the fields are back-reflected, such that there is very little flux past the interface. In the old treatment, the E fields at `z_mnt` will be interpolated from z0 and z1, and the H fields from z0.5 and z1.5. The fields will still be large and the fact that they are small past `z_interface` will not enter. In the new treatment, apart from this interpolation, an extra factor will be applied that's based on the flux at z0.5 and the flux at z1.5, with the latter depending on the fields at z2 and z2.5. Of course, in the picture above the flux at z1.5 will matter much less than the one at z0.5, but effectively what happens is a smoothening of sorts, in that the field amplitude gradually scales down as z_mnt is moved towards z_interface. Note also that once z_mnt crosses z1, the fields at z2 and z2.5 enter the old treatment, too.

It is hard for me to reason which one is a better treatment. But, to be on the safe side, maybe we want to revert to the old treatment, while *storing* the `finite_grid_correction` in the FieldData, and introduce a property like `FieldData.finite_grid_flux_copy` which returns the corrected data? In my initial discussions with @tylerflex , we agreed that we want the user to think about the finite grid as little as possible, so the goal of this PR was for the user to *not* have to call such methods/properties. However it just seems to me that there are many considerations at play, and it's hard to come up with a way to completely eliminate the grid from the user's considerations...

**Comment by momchil-flex** [comment]
In fact what I propose is that we apply the `finite_grid_flux_correction` in the `flux` and `dot` methods, such that those are still always consistent with the server-side ones, while we don't modify the raw data itself (and add a note about this in the flux and dot methods).

**Comment by tylerflex** [comment]
From a user perspective, I think the ideal case is:
1. The output from a flux monitor should not show grid-dependent oscillations when moved long its normal direction.
2. The server side and client side computed flux should be identical.
I support whatever changes are required to make this work. I don't think I understand everything going on here, but one thing that seems to be a theme is that if we have enough data in either our FieldData or ModeSolverData objects, we should be able to accomplish 1 and 2, so my overall, general comment is whether it makes sense to revert the data storage for 2D & 1D monitors so that they again store a "buffer" of data around the size=0 dimensions, instead of automatically interpolating to the normal direction. We could then provide some convenience methods to "squeeze" out the data from a 2D FieldData, for example, by inerpolating to any size=0 dimensions. But in general, would this change make it easier to get the desired behavior from your perspective?

> - Added a `ElectromagneticFieldData.time_reversed_copy` property that can be used to compute mode amplitudes in the backward direction. Note: for time-domain data, this just flips the sign of `H`. Does this make sense? If the data was a regular numpy array, I could be reversing the indexing along the time dimension. However, in the xarray data the coords are labeled, so it doesn't seem necessary?

For complex-valued data would we need to C.C. anything here? And for time coordinates, we maybe should reverse the `.coords[’t’]` instead of the array?

File: tidy3d/components/data/monitor_data.py [hunk 1]
```diff
@@ -286,6 +311,16 @@ def dot(
             and the sum of carried power fractions exceed 1. In the non-conjugated definition,
             orthogonal modes can be defined, but the interpretation of modal overlap as power
             carried by a given mode is no longer valid.
+
+        Note
```
Comment (tylerflex) [inline tidy3d/components/data/monitor_data.py:315]: it would be nice if we did the `. finite_grid_copy` automatically, if possible without the user needing to worry about it. Is it possible?

File: tidy3d/components/data/monitor_data.py [hunk 2]
```diff
@@ -310,6 +345,19 @@ def dot(
...
+        """Make a copy of the data with time-reversed fields."""
+
+        # Time reversal for frequency-domain fields; overwritten in :class:`FieldTimeData`.
+        new_data = {}
+        for comp, field in self.field_components.items():
+            if comp[0] == "H":
+                new_data[comp] = -np.conj(field)
+            else:
+                new_data[comp] = np.conj(field)
+        return self.copy(update=new_data)
```
Comment (tylerflex) [inline tidy3d/components/data/monitor_data.py:359]: perhaps we need to also reverse the `t` coords as well?

File: tidy3d/components/data/monitor_data.py [hunk 3]
```diff
@@ -434,6 +482,17 @@ def dot(self, field_data: ElectromagneticFieldData, conjugate: bool = True) -> x
         """Inner product is not defined for time-domain data."""
         raise DataError("Inner product is not defined for time-domain data.")
 
+    @property
+    def time_reversed_copy(self) -> FieldTimeData:
```
Comment (tylerflex) [inline tidy3d/components/data/monitor_data.py:486]: just a comment that this looks almost identical to the method for FieldData, except without the conj. I wonder if we should just add this method to the ElectromagenticFieldDataset more generally. Note: `np.conj(float) -> float` so can just use the same logic as FieldData here.

**Review by tylerflex — CHANGES_REQUESTED** [review]
Just a few comments, but more generally, I do think it would be nice if user didnt have to apply any grid corrections before computing flux.

File: tidy3d/components/data/monitor_data.py (hunk 3 above)
Comment (momchil-flex) [reply tidy3d/components/data/monitor_data.py:486]: Time data can be complex if Bloch boundaries are present (although we had this discussion of whether it's actually needed... but for now we return it as complex). Also with the reversal of the `t` coordinate, I guess it makes sense to keep overwriting here?

**Comment by momchil-flex** [comment]
> From a user perspective, I think the ideal case is:
> 
> 1. The output from a flux monitor should not show grid-dependent oscillations when moved long its normal direction.
> 2. The server side and client side computed flux should be identical.
> 
> I support whatever changes are required to make this work. I don't think I understand everything going on here, but one thing that seems to be a theme is that if we have enough data in either our FieldData or ModeSolverData objects, we should be able to accomplish 1 and 2, so my overall, general comment is whether it makes sense to revert the data storage for 2D & 1D monitors so that they again store a "buffer" of data around the size=0 dimensions, instead of automatically interpolating to the normal direction. We could then provide some convenience methods to "squeeze" out the data from a 2D FieldData, for example, by inerpolating to any size=0 dimensions. But in general, would this change make it easier to get the desired behavior from your perspective?

This could work, but I feel like it's a bit of an overkill in terms of how much more data will have to be stored. What I'm thinking right now is adding an optional `finite_grid_flux_correction` field to `FieldData` and `FieldTimeData`, a DataArray over `f` or `t`, respectively, which is just a factor applied during the user-side `flux` computation. Similarly, I can add a `finite_grid_e_correction` and `finite_grid_h_correction` to `ModeSolverData` already when it is returned by the solver, which would allow me to compute `.finite_grid_copy` internally without needing the user to provide a `grid`. Or maybe to unify all of those, I can do a `finite_grid_e_correction` and `finite_grid_h_correction` for all `ElectromagneticFieldData`, where for `FieldData` both of those would be just `sqrt(finite_grid_flux_correction)`. What do you think?

File: tidy3d/components/data/monitor_data.py (hunk 3 above)
Comment (tylerflex) [reply tidy3d/components/data/monitor_data.py:486]: sorry, I dont quite follow. So for complex-valued time-domain data are you saying that we dont want to apply complex conjugate for time reversal? My point was that probably we should apply np.conj to all data (including real and complex) + reverse the time coordinates and that we can likely do this with one method at the ElectromagneticFieldData level.

**Comment by tylerflex** [comment]
Personally, I think it has a better feel to not include those fields and include all of the extra data needed to make everything work out easily for client-side calculations. 

If we do want to give the user the option to save data, I might suggest we have a `interpolate_to_zero_dims: bool = False` field in `FieldMonitor` or `ModeSolverMonitor` if that makes sense. This would then return datasets that would simply not be compatible with `.flux()` or we could warn if a user calls `.flux` with eg a `FieldData` with a monitor where `interpolate_to_zero_dims` is `True`

File: tidy3d/components/data/monitor_data.py (hunk 3 above)
Comment (momchil-flex) [reply tidy3d/components/data/monitor_data.py:486]: Yeah I don't think complex time-domain data has to be conjugated. Time reversal is just something like `t -> -t`. In the frequency domain, however, that flips the sign in `exp(-1j w t)` in the Fourier transform, hence the complex conjugation.

**Comment by momchil-flex** [comment]
I think it's quite a big downside to always (or by default) have to store and download 3x the amount of data because of that small correction. Is there an argument against including the extra fields I'm suggesting that is stronger than that?

File: tidy3d/components/data/monitor_data.py (hunk 3 above)
Comment (tylerflex) [reply tidy3d/components/data/monitor_data.py:528]: ok, I would have assumed you would need to CC even in the complex-valued time domain formalism but I trust you if you say you dont. I figured the same logic applies that you could get time domain fields through FR eg `E(t) = ∫dw E(w) exp(-iwt)` and therefore would need to CC E(t) if t -> T-t

**Review by tylerflex — APPROVED** [review]
